src/paintingTools/brush.js

```diff
@@ -107,13 +107,265 @@ function onImageRendered (e) {
+  const { rows, columns } = layer.image;
+  let { x, y } = eventData.currentPoints.image;
+  const toolData = getToolState(element, TOOL_STATE_TOOL_TYPE);
+  const pixelData = toolData.data[0].pixelData;
+  const brushPixelValue = configuration.draw;
+
+  const dx = [0, 1, 0, -1];
+  const dy = [-1, 0, 1, 0];
+
+  if (x < 0 || x > columns ||
+    y < 0 || y > rows) {
+    return;
+  }
+
+  const getPixelIndex = (x, y) => (y * columns) + x;
+  const isSameColor = (x, y, color = brushPixelValue) => pixelData[getPixelIndex(x, y)] === color;
+  const draw = (x, y, brushPixelValue) => drawBrushPixels([[x, y]], pixelData, brushPixelValue, columns);
+  const drawRange = (range, color) => drawBrushPixels(range, pixelData, color, columns);
+
+  x = Math.round(x);
+  y = Math.round(y);
+
+  const find = (low, high, condition, base, isY = false) => {
+    for (let t = low; low < high ? (t <= high) : (t >= high); t += (low < high ? 1 : -1)) {
+      if (isY) {
+        if (condition(base, t)) {
+          return t;
+        }
+      } else if (condition(t, base)) {
+        return t;
+      }
+    }
+
+    return null;
+  };
+  // This thisCircle = [(x, y - radius), (x + radius, y), (x, y + radius), (x - radius, y)]
+  const thisCircle = [
+    find(y, 0, (x, y) => !isSameColor(x, y - 1), x, true),
+    find(x, columns, (x, y) => !isSameColor(x + 1, y), y, false),
+    find(y, rows, (x, y) => !isSameColor(x, y + 1), x, true),
+    find(x, 0, (x, y) => !isSameColor(x - 1, y), y, false)
+  ];
+
+  if (Math.abs(thisCircle[0] - thisCircle[2]) > radius * 2 && Math.abs(thisCircle[1] - thisCircle[3]) > radius * 2) {
+    // If the position you clicked is aleady in same color, this method is not necessary.
+    return;
+  }
+
+  // Find a direction the closest point where same color from (x, y)
+  let [minDistance, idx, base, low, high] = [2147483647, -1];
+
+  for (let i = 0; i < thisCircle.length; i += 1) {
+    let b, l, h;
+
+    if (i % 2) {
+      b = y;
+      l = find(thisCircle[i] + dx[i], i === 1 ? columns : 0, isSameColor, b, false);
+      if (l === null) {
+        continue;
+      }
+      h = find(l + dx[i], i === 1 ? columns : 0, (x, y) => !isSameColor(x + dx[i], y + dy[i]), b, false);
+    } else {
+      b = x;
+      l = find(thisCircle[i] + dy[i], i === 0 ? 0 : rows, isSameColor, b, true);
+      if (l === null) {
+        continue;
+      }
+      h = find(l + dy[i], i === 0 ? 0 : rows, (x, y) => !isSameColor(x + dx[i], y + dy[i]), b, true);
+    }
+    const distance = Math.abs(l - (i % 2 ? x : y));
+
+    if (minDistance > distance) {
+      [minDistance, idx, base, low, high] = [distance, i, b, l, h];
+    }
+  }
+
+  const getPointList = (low, high, base, isY = false) => {
+    const ret = [];
+
+    for (let t = low; low < high ? (t <= high) : (t >= high); t += (low < high ? 1 : -1)) {
+      ret.push(isY ? [base, t] : [t, base]);
+    }
+
+    return ret;
+  };
+
+
+  class Queue {
```

Comment:
I think this class should not be an inner class, and should be pulled out of the fill() function. Again, the fill() function could be easier to follow if it just contained the logical flow.

Reply:
This class is moved to queue.js